solarforecastarbiter/reports/main.py

```diff
@@ -0,0 +1,273 @@
+  2. Run validation of metadata. Metadata creation might also include
+     some validation so this may not be necessary.
+  3. Get data using io.api.
+  4. Align observation data to forecast data using metrics subpackage.
+  5. Compute metrics specified in metadata using metrics subpackage
+  6. Assemble metrics into output format. Maybe nested JSON with keys like:
+        * metadata
+            * checksum of data
+            * date created
+            * versions
+        * metrics
+            * total
+            * filter A
+                * metric 1
+                * metric 2
+            * filter B
+                * metric 1
+                * metric 2
+            * filter C
+                * metric 1
+                * metric 2
+       The JSON could be large and difficult for a person to read, but
+       should be relatively easy to parse.
+    7. Submit report to API.
+
+
+Considerations:
+
+* API uses queue system to initiate report generation
+* Functions should not require an API session unless they really need it.
+* The bokeh plots in the html version will be rendered at
+  client load time. The metrics data will be immediately available, but
+  the API will need to call the data query and realignment functions
+  to be able to create time series, scatter, etc. plots.
+"""
+
+import json
+import pkg_resources
+from pkg_resources import DistributionNotFound
+import platform
+
+import pandas as pd
+
+from solarforecastarbiter.io.api import APISession
+from solarforecastarbiter.reports import figures, template
+
+
+def get_data_for_report(session, report):
+    """
+    Get data for report.
+
+    1 API call is made for each unique forecast and observation object.
+
+    Parameters
+    ----------
+    session : solarforecastarbiter.api.APISession
+        API session for getting and posting data
+    report : solarforecastarbiter.datamodel.Report
+        Metadata describing report
+
+    Returns
+    -------
+    data : dict
+        Keys are Forecast and Observation objects, values are
+        the corresponding data.
+    """
+    data = {}
+    for fxobs in report.forecast_observations:
+        # forecasts and especially observations may be repeated.
+        # only get the raw data once.
+        if fxobs.forecast not in data:
+            data[fxobs.forecast] = session.get_forecast_values(
+                fxobs.forecast, report.start, report.end)
+        if fxobs.observation not in data:
+            data[fxobs.observation] = session.get_observation_values(
+                fxobs.observation, report.start, report.end)
+    return data
+
+
+def data_dict_to_fxobs_data(data, report):
+    """
+    Sorts the data dict into a new dict where the keys are the report's
+    ForecastObservation objects and the values are tuples of
+    (forecast values, observation values).
+    """
+    return {fxobs: (data[fxobs.forecast], data[fxobs.observation])
+            for fxobs in report.forecast_observations}
```

Comment:
does this create copies? gross

Reply:
It shouldn't if it does the normal python thing and pass by reference.